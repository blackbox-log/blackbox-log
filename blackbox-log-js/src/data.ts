import { Memoize as memoize } from 'typescript-memoize';

import { WasmPointer } from './wasm';

import type { Headers } from './headers';
import type { WasmExports, WasmObject } from './wasm';

export class DataParser implements WasmObject {
	readonly #wasm: WasmExports;
	readonly #ptr: WasmPointer;
	readonly #headers: Headers;

	constructor(wasm: WasmExports, headers: Headers, ptr: number) {
		this.#wasm = wasm;
		this.#headers = headers;
		this.#ptr = new WasmPointer(ptr, wasm.data_free);
	}

	free() {
		this.#ptr.free();
	}

	get isAlive(): boolean {
		return this.#ptr.isAlive;
	}

	get headers(): Headers {
		return this.#headers;
	}

	parseAll() {
		this.#wasm.data_parseAll(this.#ptr.ptr);
	}

	get mainFrameCount(): number {
		return this.#wasm.data_mainFrameCount(this.#ptr.ptr);
	}

	get gpsFrameCount(): number {
		return this.#wasm.data_gpsFrameCount(this.#ptr.ptr);
	}
}
