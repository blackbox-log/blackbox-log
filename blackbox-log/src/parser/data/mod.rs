mod event;

use alloc::vec::Vec;

pub use self::event::Event;
use super::{FrameKind, Headers, MainFrame, ParseError, ParseResult, Reader, SlowFrame};

#[derive(Debug, Clone)]
pub struct Data {
    pub(crate) events: Vec<Event>,
    pub(crate) main_frames: Vec<(MainFrame, usize)>,
    // pub(crate) gps_frames: Vec<Frame>,
    // pub(crate) gps_home_frames: Vec<Frame>,
    pub(crate) slow_frames: Vec<SlowFrame>,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct Stats {
    pub counts: FrameCounts,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct FrameCounts {
    pub main: usize,
    pub slow: usize,
}

impl Data {
    pub(crate) fn to_stats(&self) -> Stats {
        Stats {
            counts: FrameCounts {
                main: self.main_frames.len(),
                slow: self.slow_frames.len(),
            },
        }
    }

    pub fn parse(mut data: Reader, headers: &Headers) -> ParseResult<Self> {
        let mut events = Vec::new();
        let mut main_frames = Vec::new();
        // let gps_frames = Vec::new();
        // let gps_home_frames = Vec::new();
        let mut slow_frames = Vec::new();

        slow_frames.push(headers.slow_frames.default_frame(headers));

        let mut last_kind = None;
        while let Some(kind) = data.read_u8().map(FrameKind::from_byte) {
            // TODO (rust 1.65): let-else
            let kind = if let Some(kind) = kind {
                kind
            } else {
                tracing::error!("found invalid frame byte");
                match last_kind.take() {
                    Some(FrameKind::Event) => {
                        events.pop();
                    }
                    Some(FrameKind::Intra | FrameKind::Inter) => {
                        main_frames.pop();
                    }
                    Some(FrameKind::Slow) => {
                        slow_frames.pop();
                    }
                    Some(FrameKind::Gps | FrameKind::GpsHome) | None => {}
                };

                data.skip_until_any(
                    &[
                        FrameKind::Event,
                        FrameKind::Intra,
                        FrameKind::Slow,
                        FrameKind::Gps,
                        FrameKind::GpsHome,
                    ]
                    .map(u8::from),
                );

                continue;
            };

            let result = match kind {
                FrameKind::Event => match Event::parse_into(&mut data, &mut events) {
                    Ok(event::EventKind::End) => {
                        tracing::trace!("found the end event");
                        break;
                    }
                    Ok(_) => Ok(()),
                    Err(err) => Err(err),
                },
                FrameKind::Intra | FrameKind::Inter => {
                    let frame = MainFrame::parse(&mut data, kind, &main_frames, headers);
                    frame.map(|frame| main_frames.push((frame, slow_frames.len() - 1)))
                }
                FrameKind::Gps => {
                    if let Some(ref gps) = headers.gps_frames {
                        gps.parse(&mut data, headers).map(|_| ())
                    } else {
                        tracing::error!("found GPS frame without GPS frame definition");
                        return Err(ParseError::Corrupted);
                    }
                }
                FrameKind::GpsHome => {
                    if let Some(ref gps_home) = headers.gps_home_frames {
                        gps_home.parse(&mut data, headers).map(|_| ())
                    } else {
                        tracing::error!("found GPS home frame without GPS frame definition");
                        return Err(ParseError::Corrupted);
                    }
                }
                FrameKind::Slow => headers
                    .slow_frames
                    .parse(&mut data, headers)
                    .map(|frame| slow_frames.push(frame)),
            };

            match result {
                Ok(()) => {
                    last_kind = Some(kind);
                }
                Err(ParseError::UnexpectedEof) => {
                    tracing::warn!("found unexpected end of file");
                    break;
                }
                Err(err) => return Err(err),
            }
        }

        Ok(Self {
            events,
            main_frames,
            // gps_frames,
            // gps_home_frames,
            slow_frames,
        })
    }
}
