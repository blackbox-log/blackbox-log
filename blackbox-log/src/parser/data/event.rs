use alloc::vec::Vec;

use tracing::instrument;

use crate::parser::{decode, ParseError, ParseResult, Reader};

#[derive(Debug, Clone, PartialEq)]
pub enum Event {
    SyncBeep(u64),
    InflightAdjustment {
        function: u8,
        new_value: AdjustedValue,
    },
    Resume {
        log_iteration: u32,
        time: u32,
    },
    Disarm(u32),
    FlightMode {
        flags: u32,
        last_flags: u32,
    },
    ImuFailure {
        error: u32,
    },
    End,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum AdjustedValue {
    Float(f32),
    Int(i32),
}

impl Event {
    #[instrument(level = "debug", name = "Event::parse", skip_all, fields(kind))]
    pub(crate) fn parse_into(data: &mut Reader, events: &mut Vec<Self>) -> ParseResult<EventKind> {
        let byte = data.read_u8().ok_or(ParseError::UnexpectedEof)?;
        let kind = EventKind::from_byte(byte).unwrap_or_else(|| todo!("invalid event: {byte}"));

        match kind {
            EventKind::SyncBeep => {
                // TODO: SyncBeep handle time rollover

                let time = decode::variable(data)?;
                events.push(Self::SyncBeep(time.into()));
            }

            EventKind::InflightAdjustment => {
                let function = data.read_u8().ok_or(ParseError::UnexpectedEof)?;

                let new_value = if (function & 0x80) > 0 {
                    AdjustedValue::Float(data.read_f32().ok_or(ParseError::UnexpectedEof)?)
                } else {
                    AdjustedValue::Int(decode::variable_signed(data)?)
                };

                events.push(Self::InflightAdjustment {
                    function: function & 0x7F,
                    new_value,
                });
            }

            EventKind::Resume => {
                let log_iteration = decode::variable(data)?;
                let time = decode::variable(data)?;

                events.push(Self::Resume {
                    log_iteration,
                    time,
                });
            }

            EventKind::Disarm => {
                let reason = decode::variable(data)?;
                events.push(Self::Disarm(reason));
            }

            EventKind::FlightMode => {
                let flags = decode::variable(data)?;
                let last_flags = decode::variable(data)?;
                events.push(Self::FlightMode { flags, last_flags });
            }

            EventKind::ImuFailure => {
                let error = decode::variable(data)?;
                events.push(Self::ImuFailure { error });
            }

            EventKind::End => {
                check_message(data, b"End of log")?;

                if data.peek() == Some(b' ') {
                    // Assume INAV's new format:
                    // `End of log (disarm reason:x)\0`

                    check_message(data, b" (disarm reason:")?;

                    let reason = data.read_u8().ok_or(ParseError::UnexpectedEof)?;
                    events.push(Self::Disarm(reason.into()));

                    if data.read_u8() != Some(b')') {
                        return Err(ParseError::Corrupted);
                    }
                }

                if data.read_u8() != Some(0) {
                    return Err(ParseError::Corrupted);
                }

                events.push(Self::End);
            }
        }

        Ok(kind)
    }
}

byte_enum! {
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    #[repr(u8)]
    pub(crate) enum EventKind {
        SyncBeep = 0,
        InflightAdjustment = 13,
        Resume = 14,
        Disarm = 15,
        FlightMode = 30,
        ImuFailure = 40,
        End = 255,
    }
}

fn check_message(bytes: &mut Reader, message: &[u8]) -> ParseResult<()> {
    let bytes = bytes.read_n_bytes(message.len());

    if bytes.len() != message.len() {
        return Err(ParseError::UnexpectedEof);
    }

    if bytes != message {
        return Err(ParseError::Corrupted);
    }

    Ok(())
}
