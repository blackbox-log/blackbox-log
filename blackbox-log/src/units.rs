use alloc::vec::Vec;
use core::fmt;

use bitvec::prelude::*;

use crate::common::FirmwareKind;
use crate::parser::headers::CurrentMeterConfig;
use crate::parser::{as_signed, Headers};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Amperage {
    raw: i32,
    current_meter: CurrentMeterConfig,
}

impl Amperage {
    pub(crate) fn new(raw: u32, headers: &Headers) -> Self {
        Self {
            raw: as_signed(raw),
            current_meter: headers.current_meter.unwrap(),
        }
    }

    pub fn as_amps(&self) -> f64 {
        let milliamps = f64::from(self.raw * 3300) / 4095.;
        let milliamps = milliamps - f64::from(self.current_meter.offset);
        (milliamps * 10.) / f64::from(self.current_meter.scale)
    }
}

impl fmt::Display for Amperage {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:.*}", f.precision().unwrap_or(2), self.as_amps())
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Voltage {
    pub(crate) raw: u32,
    pub(crate) scale: u16,
}

impl Voltage {
    pub(crate) fn new(raw: u32, headers: &Headers) -> Self {
        Self {
            raw,
            scale: headers.vbat.unwrap().scale,
        }
    }

    pub fn as_volts(&self) -> f64 {
        f64::from(
            self.raw
                .saturating_mul(330)
                .saturating_mul(u32::from(self.scale)),
        ) / 4.095
    }
}

impl fmt::Display for Voltage {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:.*}", f.precision().unwrap_or(2), self.as_volts())
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Acceleration {
    raw: i32,
    one_g: u16,
}

impl Acceleration {
    pub(crate) fn new(raw: u32, headers: &Headers) -> Self {
        Self {
            raw: as_signed(raw),
            one_g: headers.acceleration_1g.unwrap(),
        }
    }

    pub fn as_gs(&self) -> f64 {
        f64::from(self.raw) / f64::from(self.one_g)
    }
}

impl fmt::Display for Acceleration {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:.*}", f.precision().unwrap_or(2), self.as_gs())
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Rotation(i32);

impl Rotation {
    pub(crate) const fn new(raw: u32) -> Self {
        Self(as_signed(raw))
    }

    pub fn as_degrees(&self) -> i32 {
        self.0
    }
}

impl fmt::Display for Rotation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

pub trait FlagSet {
    type Flag: Flag;

    fn is_set(&self, flag: Self::Flag) -> bool;
    fn as_names(&self) -> Vec<&'static str>;
}

pub trait Flag {
    fn as_name(&self) -> &'static str;
}

macro_rules! define_flag_set {
    ($set:ident, $flag_name:ident {
        $( $flag:ident : $($beta:literal)? / $($inav:literal)? ),* $(,)?
    }) => {
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
        pub struct $set {
            firmware: FirmwareKind,
            raw: BitArray<[u32; 1], Lsb0>,
        }

        impl $set {
            pub(crate) fn new(raw: u32, firmware: FirmwareKind) -> Self {
                Self {
                    firmware,
                    raw: BitArray::new([raw]),
                }
            }
        }

        impl FlagSet for $set {
            type Flag = $flag_name;

            fn is_set(&self, flag: Self::Flag) -> bool {
                flag.to_bit(self.firmware)
                    .map_or(false, |bit| self.raw[bit])
            }

            fn as_names(&self) -> Vec<&'static str> {
                self.raw
                    .iter_ones()
                    .filter_map(|bit| Some($flag_name::from_bit(bit, self.firmware)?.as_name()))
                    .collect()
            }
        }

        impl fmt::Display for $set {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                f.write_str(&self.as_names().join("|"))
            }
        }

        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
        pub enum $flag_name {
            $( $flag ),*
        }

        impl Flag for $flag_name {
            fn as_name(&self) -> &'static str {
                match self {
                    $( Self::$flag => stringify!($flag) ),*
                }
            }
        }

        impl $flag_name {
            const fn from_bit(bit: usize, firmware: FirmwareKind) -> Option<Self> {
                match (bit, firmware) {
                    $($( ($beta, FirmwareKind::Betaflight) => Some(Self::$flag), )?)*
                    $($( ($inav, FirmwareKind::INav) => Some(Self::$flag), )?)*
                    _ => None,
                }
            }

            const fn to_bit(self, firmware: FirmwareKind) -> Option<usize> {
                match (self, firmware) {
                    $($( (Self::$flag, FirmwareKind::Betaflight) => Some($beta), )?)*
                    $($( (Self::$flag, FirmwareKind::INav) => Some($inav), )?)*
                    _ => None,
                }
            }
        }

        impl fmt::Display for $flag_name {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                f.write_str(self.as_name())
            }
        }
    };
}

define_flag_set!(FlightModeSet, FlightMode {
    Angle:     1 /  0,
    Horizon:   2 /  1,
    HeadFree:  4 /  6,
    Failsafe:  6 /  9,
    Turtle:   27 / 15,

    Arm:            0 / ,
    Mag:            3 / ,
    Passthru:       5 / ,
    GpsRescue:      7 / ,
    Antigravity:    8 / ,
    HeadAdjust:     9 / ,
    CamStab:       10 / ,
    BeeperOn:      11 / ,
    LedLow:        12 / ,
    Calib:         13 / ,
    Osd:           14 / ,
    Telemetry:     15 / ,
    Servo1:        16 / ,
    Servo2:        17 / ,
    Servo3:        18 / ,
    Blackbox:      19 / ,
    Airmode:       20 / ,
    ThreeD:        21 / ,
    FpvAngleMix:   22 / ,
    BlackboxErase: 23 / ,
    Camera1:       24 / ,
    Camera2:       25 / ,
    Camera3:       26 / ,
    Prearm:        28 / ,
    BeepGpsCount:  29 / ,
    VtxPitmode:    30 / ,
    Paralyze:      31 / ,

    // User1:               32 / ,
    // User2:               33 / ,
    // User3:               34 / ,
    // User4:               35 / ,
    // PidAudio:            36 / ,
    // AcroTrainer:         37 / ,
    // VtxControlDisable:   38 / ,
    // LaunchControl:       39 / ,
    // MspOverride:         40 / ,
    // StickCommandDisable: 41 / ,
    // BeeperMute:          42 / ,

    Heading:       /  2,
    NavAltHold:    /  3,
    NavRth:        /  4,
    NavPoshold:    /  5,
    NavLaunch:     /  7,
    Manual:        /  8,
    AutoTune:      / 10,
    NavWp:         / 11,
    NavCourseHold: / 12,
    Flaperon:      / 13,
    TurnAssistant: / 14,
    Soaring:       / 16,
});

define_flag_set!(StateSet, State {
    GpsFixHome: 0 / 0,
    GpsFix:     1 / 1,
    GpsFixEver: 2 /  ,

    CalibrateMag:                 /  2,
    SmallAngle:                   /  3,
    FixedWingLegacy:              /  4,
    AntiWindup:                   /  5,
    FlaperonAvailable:            /  6,
    NavMotorStopOrIdle:           /  7,
    CompassCalibrated:            /  8,
    AccelerometerCalibrated:      /  9,
    NavCruiseBraking:             / 11,
    NavCruiseBrakingBoost:        / 12,
    NavCruiseBrakingLocked:       / 13,
    NavExtraArmingSafetyBypassed: / 14,
    AirmodeActive:                / 15,
    EscSensorEnabled:             / 16,
    Airplane:                     / 17,
    Multirotor:                   / 18,
    Rover:                        / 19,
    Boat:                         / 20,
    AltitudeControl:              / 21,
    MoveForwardOnly:              / 22,
    SetReversibleMotorsForward:   / 23,
    FwHeadingUseYaw:              / 24,
    AntiWindupDeactivated:        / 25,
    LandingDetected:              / 26,
});

define_flag_set!(FailsafePhaseSet, FailsafePhase {
    Idle:             0 / 0,
    RxLossDetected:   1 / 1,
    RxLossIdle:         / 2,
    ReturnToHome:       / 3,
    Landing:          2 / 4,
    Landed:           3 / 5,
    RxLossMonitoring: 4 / 6,
    RxLossRecovered:  5 / 7,
    GpsRescue:        6 /  ,
});
