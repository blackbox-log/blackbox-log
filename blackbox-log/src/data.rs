//! Types for the data section of blackbox logs.

use alloc::vec::Vec;

use crate::event::{self, Event};
use crate::frame::{DataFrameKind, FrameKind, GpsFrame, GpsHomeFrame, MainFrame, SlowFrame};
use crate::parser::InternalError;
use crate::{Headers, Reader};

#[derive(Debug, Clone)]
pub(crate) struct Data {
    pub(crate) events: Vec<Event>,
    pub(crate) main_frames: Vec<FrameSync>,
    pub(crate) slow_frames: Vec<SlowFrame>,
    pub(crate) gps_frames: Vec<GpsFrame>,
    pub(crate) gps_home_frames: Vec<GpsHomeFrame>,
}

#[derive(Debug, Clone)]
pub(crate) struct FrameSync {
    pub(crate) main: MainFrame,
    pub(crate) slow: usize,
    pub(crate) gps: Option<usize>,
}

impl FrameSync {
    pub(crate) const fn new(main: MainFrame, slow: &[SlowFrame], gps: &[GpsFrame]) -> Self {
        Self {
            main,
            slow: slow.len() - 1,
            gps: gps.len().checked_sub(1),
        }
    }
}

/// Statistics about a decoded log.
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[non_exhaustive]
pub struct Stats {
    /// The number of valid frames found of each type.
    pub counts: FrameCounts,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct FrameCounts {
    pub main: usize,
    pub slow: usize,
    pub gps: usize,
    pub gps_home: usize,
}

impl Data {
    pub(crate) fn to_stats(&self) -> Stats {
        Stats {
            counts: FrameCounts {
                main: self.main_frames.len(),
                slow: self.slow_frames.len(),
                gps: self.gps_frames.len(),
                gps_home: self.gps_home_frames.len(),
            },
        }
    }

    pub(crate) fn parse(data: &mut Reader, headers: &Headers) -> Self {
        let mut events = Vec::new();
        let mut main_frames = Vec::new();
        let mut slow_frames = Vec::new();
        let mut gps_frames = Vec::new();
        let mut gps_home_frames = Vec::new();

        slow_frames.push(headers.slow_frames.empty_frame());

        if let Some(def) = &headers.gps_frames {
            gps_frames.push(def.empty_frame());
        }

        let mut restore;
        let mut last_kind = None;
        while let Some(byte) = data.read_u8() {
            restore = data.get_restore_point();

            let Some(kind) = FrameKind::from_byte(byte) else {
                tracing::debug!("found invalid frame byte: {byte:0>#2x}");

                if let Some(last_kind) = last_kind.take() {
                    data.restore(restore);

                    match last_kind {
                        FrameKind::Event => {
                            events.pop();
                        }
                        FrameKind::Data(DataFrameKind::Intra | DataFrameKind::Inter) => {
                            main_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::Slow) => {
                            slow_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::Gps) => {
                            gps_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::GpsHome) => {
                            gps_home_frames.pop();
                        }
                    };
                }

                skip_to_frame(data);
                continue;
            };

            tracing::trace!("trying to parse {kind:?} frame");

            let result = match kind {
                FrameKind::Event => match Event::parse_into(data, &mut events) {
                    Ok(event::EventKind::End) => {
                        tracing::trace!("found the end event");
                        break;
                    }
                    Ok(_) => Ok(()),
                    Err(err) => Err(err),
                },
                FrameKind::Data(DataFrameKind::Intra | DataFrameKind::Inter) => {
                    let frame = MainFrame::parse(data, kind, &main_frames, headers);
                    frame.map(|frame| {
                        main_frames.push(FrameSync::new(frame, &slow_frames, &gps_frames));
                    })
                }
                FrameKind::Data(DataFrameKind::Gps) => headers.gps_frames.as_ref().map_or_else(
                    || {
                        tracing::debug!("found GPS frame without GPS frame definition");
                        Err(InternalError::Retry)
                    },
                    |gps| {
                        gps.parse(
                            data,
                            headers,
                            main_frames.last().map(|sync| &sync.main),
                            gps_home_frames.last(),
                        )
                        .map(|frame| gps_frames.push(frame))
                    },
                ),
                FrameKind::Data(DataFrameKind::GpsHome) => {
                    headers.gps_home_frames.as_ref().map_or_else(
                        || {
                            tracing::debug!(
                                "found GPS home frame without GPS home frame definition"
                            );
                            Err(InternalError::Retry)
                        },
                        |gps_home| {
                            gps_home
                                .parse(data, headers)
                                .map(|frame| gps_home_frames.push(frame))
                        },
                    )
                }
                FrameKind::Data(DataFrameKind::Slow) => headers
                    .slow_frames
                    .parse(data, headers)
                    .map(|frame| slow_frames.push(frame)),
            };

            match result {
                Ok(()) => {
                    last_kind = Some(kind);
                }
                Err(InternalError::Retry) => {
                    tracing::debug!("found corrupted {kind:?} frame");
                    data.restore(restore);
                    skip_to_frame(data);
                }
                Err(InternalError::Eof) => {
                    tracing::debug!("found unexpected end of file in data section");
                    break;
                }
            }
        }

        Self {
            events,
            main_frames,
            slow_frames,
            gps_frames,
            gps_home_frames,
        }
    }
}

#[cold]
fn skip_to_frame(data: &mut Reader) {
    data.skip_until_any(
        &[
            FrameKind::Event,
            FrameKind::Data(DataFrameKind::Intra),
            FrameKind::Data(DataFrameKind::Slow),
            FrameKind::Data(DataFrameKind::Gps),
            FrameKind::Data(DataFrameKind::GpsHome),
        ]
        .map(u8::from),
    );
}
