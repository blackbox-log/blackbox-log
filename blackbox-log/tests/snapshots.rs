use std::collections::BTreeMap;
use std::fs;
use std::io::Read;

use blackbox_log::data::{ParseEvent, Stats};
use blackbox_log::event::Event;
use blackbox_log::units::{si, Flag, FlagSet};
use blackbox_log::{DataParser, Headers, Unit, Value};
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

macro_rules! run {
    () => {
        |path| {
            let mut file = fs::File::open(path).unwrap();
            let mut data = Vec::new();
            file.read_to_end(&mut data).unwrap();

            let file = blackbox_log::File::new(&data);
            let logs = file
                .iter()
                .map(|mut reader| {
                    Headers::parse(&mut reader).map(|headers| {
                        let data = DataParser::new(&mut reader, &headers);
                        LogSnapshot::new(&headers, data)
                    })
                })
                .collect::<Vec<_>>();

            insta::assert_ron_snapshot!(logs);
        }
    };
}

#[test]
fn own() {
    insta::glob!("logs/*.bbl", run!());
}

#[test]
#[ignore]
fn fc_blackbox() {
    insta::glob!("logs/fc-blackbox/*", run!());
}

#[test]
#[ignore]
fn gimbal_ghost() {
    insta::glob!("logs/gimbal-ghost/*", run!());
}

#[derive(Debug, Serialize)]
struct LogSnapshot<'data> {
    headers: Headers<'data>,
    stats: Stats,
    events: Vec<Event>,
    main: Fields,
    slow: Fields,
    gps: Fields,
}

impl<'data> LogSnapshot<'data> {
    fn new(headers: &Headers<'data>, mut data: DataParser<'data, '_, '_>) -> Self {
        let headers = headers.clone();

        let mut events = Vec::new();
        let mut main = headers.main_def().iter().collect::<Fields>();
        let mut slow = headers.slow_def().iter().collect::<Fields>();
        let mut gps = headers
            .gps_def()
            .iter()
            .flat_map(|def| def.iter())
            .collect::<Fields>();

        while let Some(frame) = data.next() {
            match frame {
                ParseEvent::Event(event) => events.push(event),
                ParseEvent::Main(frame) => main.update(frame),
                ParseEvent::Slow(frame) => slow.update(frame),
                ParseEvent::Gps(frame) => gps.update(frame),
            }
        }

        Self {
            headers,
            stats: data.stats().clone(),
            events,
            main,
            slow,
            gps,
        }
    }
}

#[derive(Debug, Serialize)]
struct Fields(Vec<FieldSnapshot>);

impl Fields {
    fn update<F: blackbox_log::frame::Frame>(&mut self, frame: F) {
        for (field, value) in self.0.iter_mut().zip(frame.iter()) {
            field.update(value.into());
        }
    }
}

impl<T, U> FromIterator<(T, U)> for Fields
where
    T: Into<String>,
    U: Into<Unit>,
{
    fn from_iter<I: IntoIterator<Item = (T, U)>>(iter: I) -> Self {
        Self(
            iter.into_iter()
                .map(|(name, unit)| FieldSnapshot::new(name.into(), unit.into()))
                .collect(),
        )
    }
}

#[derive(Debug, Serialize)]
struct FieldSnapshot {
    name: String,
    unit: Unit,
    history: History,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum History {
    Int(NumberHistory<16, i128>),
    Bool { yes: usize, no: usize },
    Flags(BTreeMap<&'static str, usize>),
}

#[derive(Debug)]
struct NumberHistory<const N: usize, T> {
    min: T,
    max: T,
    seen: Vec<T>,
    histogram: [usize; N],
}

impl FieldSnapshot {
    fn new(name: String, unit: Unit) -> Self {
        Self {
            name,
            unit,
            history: match unit {
                Unit::FrameTime
                | Unit::Amperage
                | Unit::Voltage
                | Unit::Acceleration
                | Unit::Rotation
                | Unit::GpsCoordinate
                | Unit::Altitude
                | Unit::Velocity
                | Unit::GpsHeading
                | Unit::Unitless => History::Int(NumberHistory::new()),
                Unit::Boolean => History::Bool { yes: 0, no: 0 },
                Unit::FlightMode | Unit::State | Unit::FailsafePhase => {
                    History::Flags(BTreeMap::new())
                }
            },
        }
    }

    #[allow(clippy::wildcard_enum_match_arm, clippy::cast_possible_truncation)]
    fn update(&mut self, value: Value) {
        match &mut self.history {
            History::Int(history) => history.update(match value {
                Value::FrameTime(t) => t.get::<si::time::microsecond>().round() as i128,
                Value::Amperage(a) => a.get::<si::electric_current::milliampere>().round() as i128,
                Value::Voltage(v) => v.get::<si::electric_potential::millivolt>().round() as i128,
                Value::Acceleration(a) => a
                    .get::<si::acceleration::centimeter_per_second_squared>()
                    .round() as i128,
                Value::Rotation(r) => {
                    r.get::<si::angular_velocity::degree_per_second>().round() as i128
                }
                Value::GpsCoordinate(c) => (c * 10000000.).round() as i128,
                Value::Altitude(a) => a.get::<si::length::meter>().round() as i128,
                Value::Velocity(v) => {
                    v.get::<si::velocity::centimeter_per_second>().round() as i128
                }
                Value::GpsHeading(h) => (h * 10.).round() as i128,
                Value::Unsigned(u) => u.into(),
                Value::Signed(s) => s.into(),
                _ => unreachable!(),
            }),
            History::Bool { yes, no } => match value {
                Value::Boolean(true) => *yes += 1,
                Value::Boolean(false) => *no += 1,
                _ => unreachable!(),
            },
            History::Flags(history) => {
                if let Value::FailsafePhase(phase) = value {
                    *history.entry(phase.as_name()).or_insert(0) += 1;
                } else {
                    let flags = match value {
                        Value::FlightMode(m) => m.as_names(),
                        Value::State(s) => s.as_names(),
                        _ => unreachable!(),
                    };

                    for flag in flags {
                        *history.entry(flag).or_insert(0) += 1;
                    }
                }
            }
        }
    }
}

impl<const N: usize, T> NumberHistory<N, T> {
    fn new() -> Self
    where
        T: Default,
    {
        Self {
            min: T::default(),
            max: T::default(),
            seen: Vec::new(),
            histogram: [0; N],
        }
    }

    fn update_range(&mut self, value: T)
    where
        T: PartialOrd,
    {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }
    }

    /// Store new value sorted in `seen` if not already present
    fn update_seen(&mut self, value: T)
    where
        T: PartialOrd,
    {
        let index = self.seen.partition_point(|x| *x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }
    }

    /// Bucket so that sequential values go in sequential buckets, since data is
    /// usually clustered
    #[inline(always)]
    fn update_histogram(&mut self, value: u128) {
        self.histogram[(value % N as u128) as usize] += 1;
    }
}

impl<const N: usize> NumberHistory<N, i128> {
    fn update(&mut self, value: i128) {
        self.update_range(value);
        self.update_seen(value);
        self.update_histogram(value.unsigned_abs());
    }
}

impl<const N: usize> Serialize for NumberHistory<N, i128>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(self, self.min, self.max, serializer)
    }
}

impl<const N: usize> Serialize for NumberHistory<N, f64>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(
            self,
            format!("{:.2}", self.min),
            format!("{:.2}", self.max),
            serializer,
        )
    }
}

#[inline(always)]
fn serialize<const N: usize, T, S, U>(
    snapshot: &NumberHistory<N, T>,
    min: U,
    max: U,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    U: Serialize,
    [usize; N]: Serialize,
{
    let unique = snapshot.seen.len();

    let mut state = serializer.serialize_struct("FieldSnapshot", 4)?;

    state.serialize_field("min", &min)?;
    state.serialize_field("max", &max)?;
    state.serialize_field("unique", &unique)?;

    let histogram = "histogram";
    if unique > 1 {
        state.serialize_field(histogram, &snapshot.histogram)?;
    } else {
        state.skip_field(histogram)?;
    }

    state.end()
}
