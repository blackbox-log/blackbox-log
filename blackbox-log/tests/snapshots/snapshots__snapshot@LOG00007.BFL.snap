---
source: blackbox-log/tests/snapshots.rs
expression: logs
input_file: blackbox-log/tests/logs/LOG00007.BFL
---
[
  Err(Corrupted),
]
